Initial draft of Homebrew migration guide (#96)
flox/floxdocs: docs/tutorials/migrations/homebrew.md

---
title: Homebrew
description: Common questions and solutions for using Rust with Flox
---

# Homebrew migration guide

> Flox is a virtual environment manager and package manager all in one. Using Flox, you can create virtual environments containing packages from the Flox Catalog.

Reviewer: The word "virtual" isn't something I usually say when talking about Flox Environments (nor do our taglines). Is it here because of the familiarity some folks have with Python venvs?

Reviewer 2: We say "Flox is a virtual environment and package manager all in one." on the home page of the docs, but not "virtual environment **manager**"

* **You need reproducible environments.** Flox environments are defined in a TOML manifest that can be easily shared, or managed alongside code.
* **You need a broader set of software.** Homebrew has a sizeable collection of packages, but the Flox Catalog is based on Nixpkgs, the largest collection of packages in the world.
* **You need older versions of software.** The Flox Catalog contains historical versions of each of its packages - in many cases, going back decades.

## Install Flox
Download and install Flox following the [installation instructions]().

## Find and install your first package
Migrating to Flox is a straightforward process of installing Flox packages for each of your Homebrew formulae. This section walks you through the process for identifying the set of Homebrew formulae you currently have installed, searching for the corresponding packages in the Flox Catalog, and installing them.

As a Homebrew user, you will find several of the Flox subcommands familiar:

* `search` is used to find available packages
* `install` is used to install packages
* `uninstall` is used to remove packages

> ### Show top-level formulae in Homebrew

Reviewer: We don't mention how to migrate if you have a `Brewfile`.  That might be a really small note, but I haven't played with them too much.

Author: Good point. I think we should do a bit more investigation and add this as a second pass.

First, identify the list of Homebrew packages you have installed.

We recommend using `brew leaves` for this, so you can easily differentiate between the formulae you installed explicitly versus those that were installed as dependencies. The `leaves` subcommand will show the formulae that were directly installed.

For example:

```
% brew leaves
boost
gh
imagemagick
mysql
openssh
```

Then, search Flox for one of the formulae you have installed with Homebrew. In this case, for example, you could choose `boost`:

```

> % flox search boost

Reviewer: This example is odd in that there are many packages at concurent versions on each revision of nixpkgs (so versioning happens via the name and via the version). 

This is totally a nonblocking suggestion, but I wonder if another package would be easier to illustrate the points with.

boost     Collection of C++ libraries
booster   Fast and secure initramfs generator
xgboost   Scalable, Portable and Distributed Gradient Boosting (GBDT, GBRT or GBM) Library
boost155  Collection of C++ libraries
boost159  Collection of C++ libraries
boost15x  Collection of C++ libraries
boost160  Collection of C++ libraries
boost165  Collection of C++ libraries
boost166  Collection of C++ libraries
boost167  Collection of C++ libraries

Showing 10 of 120 results. Use `flox search boost --all` to see the full list.

Use 'flox show <package>' to see available versions
```

The first one on the list is the correct Flox package to install, and it has the same name as the Homebrew package. This will often be the case, but not always.

### Install your first package
To install your first package, use `flox install`:

```
% flox install boost
```

The first time you install a package, Flox will ask you whether you want to create a [default environment](). If you choose to do so, Flox will then ask you whether you want to configure your shell to automatically activate the new default environment.

```

> [show the flow]

Reviewer: I assume this is a placeholder :)

```

When Flox is configured with a default environment, it behaves very similarly to Homebrew. The Flox CLI will assume the default environment when you run `flox install` in a directory that doesn't contain an environment of its own.

Creating a Flox default environment is optional.

If you do not choose for this to be automated at the time of your first package installation, you can [follow these instructions]() to add Flox to your dotfiles manually.

### Verifying configuration

> Exit your active shell and create a new one, causing the dotfile changes to take effect. The first time this happens, you may experience a short delay while your default environment is materialized. The next time you open a shell it should be quick because the environment has been cached.

Reviewer: ~~short~~

### Install your first package
To install your first package, use `flox install`:

```
% flox install boost
```

The first time you install a package, Flox will ask you whether you want to create a [default environment][default-tutorial]. If you choose to do so, Flox will then ask you whether you want to configure your shell to automatically activate the new default environment.

```
[show the flow]
```

When Flox is configured with a default environment, it behaves very similarly to Homebrew. The Flox CLI will assume the default environment when you run `flox install` in a directory that doesn't contain an environment of its own.

Creating a Flox default environment is optional.

If you do not choose for this to be automated at the time of your first package installation, you can [follow these instructions][default-tutorial#setup] to add Flox to your dotfiles manually.

### Verify configuration
Exit your active shell and create a new one, causing the dotfile changes to take effect. The first time this happens, you may experience a short delay while your default environment is materialized. The next time you open a shell it should be quick because the environment has been cached.

> Once the shell is available, you should see `flox [default]` prepended to your shell prompt. You can also verify that your default environment is active by running [`flox envs`][envs]:

Reviewer 2: I thought we suppress the prompt for the default env?

Author: Agh! you are correct, let me revert that back.

Flox can replace [Homebrew](https://brew.sh) entirely, or they can be used together.

This guide explains how to introduce Flox into environments where Homebrew is currently being used, either as a replacement or an addition. It introduces new concepts and proposes a basic procedure for mapping packages.

## Why you might want to migrate
Homebrew does a great job, and has been loved as the “missing package manager” for a generation of macOS users. But there are a few reasons you might consider moving to Flox:

* **You need a cross-platform package manager.** Flox works on both macOS and Linux - x86 and ARM - allowing you to define a set of packages that works the same everywhere.
* **You need a virtual package manager** Flox allows you to create as many environments as you need, each with a different set of packages.
* **You need reproducible environments.** Flox environments are defined in a TOML [manifest][manifest_concept] that can be easily shared, or managed alongside code.
* **You need a broader set of software.** Homebrew has a sizeable collection of packages, but the Flox Catalog is based on Nixpkgs, the largest collection of packages in the world.

> * **You need older versions of software.** The Flox Catalog contains historical versions of each of its packages - in many cases, going back decades.

Reviewer 3: > going back decades

I don't think that's true. Our catalog only goes back about 3 years as far as I'm aware. The Nix binary cache may have entries that old, but they aren't really accessible through our catalog.

> * **You need older versions of software.** The Flox Catalog contains historical versions of each of its packages - in many cases, going back decades.

Author: ```suggestion
* **You need older versions of software.** The Flox Catalog contains historical versions of each of its packages, and makes it easy to install them.
```

As a Homebrew user, you will find several of the Flox subcommands familiar:

* [`search`][search] is used to find available packages
* [`install`][install] is used to install packages
* [`uninstall`][uninstall] is used to remove packages

### Show top-level formulae in Homebrew
First, identify the list of Homebrew packages you have installed.

We recommend using `brew leaves` for this, so you can easily differentiate between the formulae you installed explicitly versus those that were installed as dependencies. The `leaves` subcommand will show the formulae that were directly installed.

For example:

```
% brew leaves
aider
awscli
ffmpeg
flake8
gh
gnupg
go
graphviz
htop
isort
jq
neovim
pkgconf
redo
ripgrep
tmux
tree
watch
wget
```

### Search for a package in Flox
Then, search Flox for one of the formulae you have installed with Homebrew. In this case, for example, you could choose `jq`:

```

> % fflox search jq

Reviewer 2: typo: fflox

Author: ```suggestion
% flox search jq
```
